dsp-user/handler/src/main/scala/dsp/user/handler/UserHandler.scala

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright © 2021 - 2022 Swiss National Data and Service Center for the Humanities and/or DaSCH Service Platform contributors.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package dsp.user.handler
+
+import dsp.user.api.UserRepo
+import zio._
+import java.util.UUID
+import dsp.errors.BadRequestException
+import dsp.errors.NotFoundException
+import dsp.errors.DuplicateValueException
+import dsp.user.domain.User
+import dsp.user.domain.UserId
+import dsp.valueobjects.User._
+import dsp.errors.RequestRejectedException
+
+/**
+ * The user handler.
+ *
+ * @param repo  the user repository
+ */
+final case class UserHandler(repo: UserRepo) {
+  // implement all possible requests from V2, but divide things up into smaller functions to keep it cleaner than before
+
+  /**
+   * Retrieve all users (sorted by IRI).
+   */
+  def getUsers(): UIO[List[User]] =
+    repo.getUsers().map(_.sorted)
+
+  // getSingleUserADM should be inspected in the route. According to the user identifier, the
+  // right method from the userHandler should be called.
+
+  /**
+   * Retrieve the user by ID.
+   *
+   * @param id  the user's ID
+   */
+  def getUserById(id: UserId): IO[NotFoundException, User] =
+    for {
+      user <- repo.getUserById(id).mapError(_ => NotFoundException("User not found"))
+    } yield user
+
+  /**
+   * Retrieve the user by username.
+   *
+   * @param username  the user's username
+   */
+  def getUserByUsername(username: Username): IO[NotFoundException, User] =
+    repo
+      .getUserByUsernameOrEmail(username.value)
+      .mapError(_ => NotFoundException(s"User with Username ${username.value} not found"))
+
+  /**
+   * Retrieve the user by email.
+   *
+   * @param email  the user's email
+   */
+  def getUserByEmail(email: Email): IO[NotFoundException, User] =
+    repo
+      .getUserByUsernameOrEmail(email.value)
+      .mapError(_ => NotFoundException(s"User with Email ${email.value} not found"))
+
+  /**
+   * Check if username is already taken
+   *
+   * @param username  the user's username
+   */
+  private def checkUsernameTaken(username: Username): IO[DuplicateValueException, Unit] =
+    for {
+      _ <- repo
```

Comment:
this looks a bit ugly. IMHO it would be better to add a more fitting method to the repo.